cjdns-core/src/announcement/serialized_ann.rs

```diff
@@ -0,0 +1,744 @@
+        Ok((node_encryption_key, node_ip6))
+    }
+
+    fn parse_entities(entities_data: &[u8]) -> Result<AnnouncementEntities> {
+        let mut parsed_entities = Vec::new();
+        let mut idx = 0_usize;
+        while idx < entities_data.len() {
+            let encoded_entity_len = entities_data[idx] as usize;
+            let entity_data = entities_data
+                .get(idx..idx + encoded_entity_len)
+                .ok_or(ParserError::CannotParseEntity("entity with invalid encoded length"))?;
+
+            if encoded_entity_len == 0 {
+                return Err(ParserError::CannotParseEntity("zero entity inside entities data"));
+            }
+            if encoded_entity_len == 1 {
+                idx += 1;
+                continue;
+            }
+
+            let &entity_type = entities_data.get(idx + 1).expect("entity data length is ne to encoded length in it");
+            let parsed_entity = parse_entity(entity_type, entity_data)?;
+            if let Some(entity) = parsed_entity {
+                parsed_entities.push(entity)
+            }
+
+            idx += entities_data[idx] as usize;
+        }
+
+        Ok(parsed_entities)
+    }
+
+    fn parse_entity(entity_type: u8, entity_data: &[u8]) -> Result<Option<Entity>> {
+        // First byte of each entity data is its length. The second byte is its type. So "non meta" data of each entity starts from index 2 (3d byte).
+        let parsing_data = &entity_data[2_usize..];
+        match entity_type {
+            ENCODING_SCHEME_TYPE => {
+                if entity_data.len() < ENCODING_SCHEME_ENTITY_MIN_SIZE || entity_data.len() > ENTITY_MAX_SIZE {
+                    return Err(ParserError::CannotParseEntity("invalid encoding scheme data size"));
+                }
+                let scheme_entity = parse_encoding_scheme(parsing_data)?;
+                Ok(Some(scheme_entity))
+            }
+            PEER_TYPE => {
+                if entity_data.len() != PEER_ENTITY_SIZE {
+                    return Err(ParserError::CannotParseEntity("invalid peer data size"));
+                }
+                let peer_entity = parse_peer(parsing_data)?;
+                Ok(Some(peer_entity))
+            }
+            VERSION_TYPE => {
+                if entity_data.len() != VERSION_ENTITY_SIZE {
+                    return Err(ParserError::CannotParseEntity("invalid version data size"));
+                }
+                let version_entity = parse_version(parsing_data)?;
+                Ok(Some(version_entity))
+            }
+            LINK_STATE_TYPE => {
+                if entity_data.len() < LINK_STATE_ENTITY_MIN_SIZE || entity_data.len() > ENTITY_MAX_SIZE {
+                    return Err(ParserError::CannotParseEntity("invalid link state data size"));
+                }
+                let link_state_entity = parse_link_state(parsing_data)?;
+                Ok(Some(link_state_entity))
+            }
+            _ => Ok(None),
```

Comment:
Maybe add some `Unknown(type: u8)` entity? Would it be helpful? (I do not insist, though).